### DISCUSSION TITLE
Dot-sourcing common build script for multiple subprojects and parameters

### DISCUSSION THREAD

**Discussion by jperterm** [nightroman/Invoke-Build#223] [General]
Hi, 

first of all: Invoke-Build is a really simple to use and straight-forward tool for a lot of build tasks, thanks a lot!

Now, I have a (maybe) special use-case in which I want to re-use a common build script in a repository with a lot of subprojects. Those subprojects are of similar type, but with different configurations. I manage these configurations with JSON files.

To illustrate by example:

Project structure:

```
- common
  - common.build.ps1
- subproject1
  - subproject.build.ps1
- subproject2
- ...
- subprojectn
```

Sample content of `common.build.ps1`:

```pwsh
param (
    [string]$TestCase
)

task Debug {
    Write-Output "You're debugging test-case '$TestCase'"
}
```

And here the content of `subproject1.build.ps1`:

```pwsh
. '../common/common.build.ps1'
```

As you can see, I am dot-sourcing the `common.build.ps1` into the subproject build file. This allows me to make improvements to the common build script, without making copies again and again, and also managing different versions.

To come to the point/problem:

Dot-sourcing the common script into the subproject works well with tasks. I can do `Invoke-Build Debug` and `Invoke-Build ?` and I get no different behaviour as when having a full copy of the build script in each subproject.

But when I want to pass parameters like `Invoke-Build Debug -TestCase test1`, then I get

```
Invoke-Build.ps1: A parameter cannot be found that matches parameter name 'TestCase'.
```

This behaviour somehow makes sense when I compare with "normal" powershell scripts (not called via `Invoke-Build`). Every script has its own parameters and indeed the subproject build files have no parameters at all.

Do you (or anyone) know what could be a best-practise for my use-case? I still want to be able to call `Invoke-Build` in the subproject.

**Comment by nightroman** [comment]
@jperterm Thank you. It's an interesting use case and the issue. Let me play/think a little bit.

**Comment by nightroman** [comment]
**One possible approach**
(note amended project structure and use of `*.tasks.ps1` instead of `*build.ps1` for subprojects)

Project structure:

```
- common.build.ps1
- subproject1
  - subproject1.tasks.ps1
- subproject2
  - subproject2.tasks.ps1
- ...
```

Sample content of `common.build.ps1`:

```powershell
param (
    [string]$TestCase
)

# code before
#...

. $TestCase\$TestCase.tasks.ps1

# code after
#...
```

And here the content of `subproject1.tasks.ps1`, just own tasks, do not call `common.build.ps1`, + count on some `# code before` invoked for this script + count on common parameters:

```powershell
task t1 {...}
...
```

Having done this, calling `Invoke-Build Debug -TestCase test1` should work in the root folder and any subfolder.

**Comment by nightroman** [comment]
FWIW, I have created the demo: [SubCases](https://github.com/nightroman/Invoke-Build/tree/main/Tasks/SubCases)

**Comment by jperterm** [comment]
Thank you for your immediate feedback! I'll try that out and give feedback here soon.

**Comment by jperterm** [comment]
I checked your proposal, but it does not really fit for my use-case. I guess I somehow set you on the wrong track with my parameter example (`$TestCase`), or maybe I simply do not understand it right.

I understand that `Invoke-Build` will find the `common.build.ps1` in the parent dir. And with the help of the parameter, the subproject's build script can be dot-sourced, and add tasks, etc. I will try to explain my view:

In your documentation you mention this:

> **$BuildRoot**
>
> It is the full path of the build script directory, by default. Build scripts may alter it on loading in special cases. For example:
>
> - It is not suitable to keep a build script in a directory where it builds.
> - A build script is designed to work for several or variable directories.

The last sentence "A build script is designed to work for several or variable directories." matches my use-case best.

Plus I want to be able to have the `common.build.ps1` in multiple major versions. This can be neccessary, when breaking changes occur in the build pipeline (which I want to avoid as long as possible). Then I must be able to upgrade a subproject A to `common.2.build.ps1` before subproject B which must still be build with `common.1.build.ps1`.

```
- common
  - common.1.build.ps1
  - common.2.build.ps1
...
```

Finally, as mentioned I landed in your docs about the `$BuildRoot` variable. I come nearest to what I want, when I call the subproject's build file directly (e.g. `./subproject1.build.ps1`) with such contents:

```pwsh
Invoke-Build @args -File '..\common\common.1.build.ps1' -BuildRoot $PSScriptRoot
```

So what I loose in the end is, that I cannot type `ib` or `Invoke-Build` in the console (which I am used to and also liking a lot) like in my other projects.

Nonetheless, thank you for your help!

UPDATE: The example for the `subproject1.build.ps1` I give above is meant to be a full sample. It shall not contain any other tasks or script logic.

**Comment by nightroman** [comment]
Yes, I see that your use case is more involved than my `SubCases`, especially with the requirement of supporting several versions of the main build script (interesting technique, by the way).

If you do not mind using dynamic parameters in subproject scripts (with a common helper, to avoid same boilerplate), this might be the way to explore.

UPDATE: I've tried dynamic parameters. So far even with a helper it looks not so good, too much cryptic code noise.

**Comment by nightroman** [comment]
@jperterm Some new development, might be related to your use case.
[New feature: Build script inheritance · nightroman/Invoke-Build · Discussion #227](https://github.com/nightroman/Invoke-Build/discussions/227)

**Comment by nightroman** [comment]
I think that build script inheritance based of special parameters `Extends` covers your case exactly, use the latest IB v5.14.3

Namely, in your `subproject1.build.ps1` instead of explicit dot-sourcing, use the special parameter `Extends`:

```powershell
param(
	[ValidateScript({'../common/common.build.ps1'})]
	$Extends
)
```

And this is it.

Now in the `subproject1.build.ps1` directory, invoke, as before:

```powershell
Invoke-Build Debug -TestCase test1
```

and instead of the error get this expected result

```
Build Debug ...\2025-05-28-2246-ib-223\subproject1\subproject.build.ps1
Task /Debug
You're debugging test-case 'test1'
Done /Debug 00:00:00.0113951
Build succeeded. 1 tasks, 0 errors, 0 warnings 00:00:00.1966942
```

Here it is ready to play:
[2025-05-28-2246-ib-223.zip](https://github.com/user-attachments/files/20495583/2025-05-28-2246-ib-223.zip)

**Reply by jperterm** [reply]
Got it working now.

As all given paths are relative to the extending scripts directory, I needed to add this in my extended script (aka "common"):

```pwsh
Set-Location $OriginalLocation
$BuildRoot = '.'
```

I wonder if there is another way of doing this from the extending script, but actually I see no significant reason. Everything seems to work fine this way.

**Reply by nightroman** [reply]
Altering `$BuildRoot` is very much expected and totally normal and gives you control.
Especially now with 2+ kind of isolated scripts playing the game of extending.

**Reply by nightroman** [reply]
I am very glad that the new feature works well for you. It works for me well, too.
Extending scripts seems to be a powerful tool already (formally preview).

**Reply by nightroman** [reply]
A comment and a concern...

**(1) FWIW, this "boilerplate"**

```
Set-Location $OriginalLocation
$BuildRoot = '.'
```

can be one-liner

```
Set-Location ($BuildRoot = $OriginalLocation)
```

**(2) I have doubts, depending on the design**

If the design of `common` is to work at the `subproject` directory, then we do not always have it done.

See attached and run
- `test1-call-subproject1-there.ps1`
    - get expected -- `Enter common @ C:\tmp\2025-05-28-2246-ib-223-2\subproject1`
- `test2-call-subproject1-here.ps1`
    - potentially unexpected -- `Enter common @ C:\tmp\2025-05-28-2246-ib-223-2`

[2025-05-28-2246-ib-223-2.zip](https://github.com/user-attachments/files/20622642/2025-05-28-2246-ib-223-2.zip)

**Reply by nightroman** [reply]
I think IB should provide an extra variable for getting this done properly.

To cover multilevel `Extends`, we need some array `$BuildRoots` in addition to `$BuildRoot`, where
- `$BuildRoots[0]` is `$BuildRoot`.
- `$BuildRoots[1]` is the direct `Extends` script root, if any (!).
- ...
- `$BuildRoots[-1]` is the root in the whole `Extends` tree, including no tree, just `common`, i.e. always valid.

Then `common` should use one of `$BuildRoots[1] .. $BuildRoots[-1]`, depending on design (in our case `[1]` and `[-1]` are the same)

```
Set-Location ($BuildRoot = $BuildRoots[-1])
```

I have it done locally and it seems to work well, fixing the current issue.
Thoughts?

**Reply by jperterm** [reply]
(1) Thanks for the tip. Besides being a one-liner -- your suggestion is more declarative.

(2) You are absolutely right. `$OriginalLocation` of course *is not* the location of the top-level (extending) build script ... I did not even think of it.

My first thought was to have an extra variable just for the top-level (-1). But your proposal seems useful to me. Even if for my use case the top-level script is the one I am interested in. I suppose it is a good idea to consider use cases where you have more than two levels.

**Reply by nightroman** [reply]
Great, thank you. After some more testing it will be in, likely today.

**Reply by jperterm** [reply]
Least I can do, is giving you some feedback. So thanks from my side again.

**Reply by nightroman** [reply]
It is in, v5.14.8

**Reply by nightroman** [reply]
One more comment, "tips and tricks", just FYI

```
Set-Location ($BuildRoot = $BuildRoots[-1])
```

Strictly speaking, `Set-Location` is redundant here. After loading the script IB keeps `$BuildRoot` (original or altered by the script) and then uses it for `Set-Location $BuildRoot` before invoking any user code from this script.

So this is enough in most cases:

```
$BuildRoot = $BuildRoots[-1]
```

Unless of course the changed location is actually used by the script during loading. This is possible but ideally should be avoided, i.e. such code moved to `Enter-Build` block from the top level script code, as a recommended practice.

**Reply by jperterm** [reply]
I installed v5.14.9 and set only `$BuildRoot` like you recommended. This forced me to move few initialisation variables to Enter-Build (which I need anyway). So for me, what to keep in mind is "always resolve paths within tasks".

**Comment by jperterm** [comment]
Wow, thanks a lot for the update on this topic and of course -- for your work!

It is still an issue for me. I will try out soon and report back!
